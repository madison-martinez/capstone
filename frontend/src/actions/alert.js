export const alertActionTypes = {
    SUCCESS: 'ALERT_SUCCESS',
    ERROR: 'ALERT_ERROR',
    RESET: 'ALERT_RESET'
};

function success(message) {
    return {
        type: alertActionTypes.SUCCESS,
        message
    }
};

function error(message) {
    return {
        type: alertActionTypes.ERROR,
        message
    }
};

function reset() {
    return {
        type: alertActionTypes.RESET
    }
};


export const alertActions = {
    success,
    error,
    reset
};

