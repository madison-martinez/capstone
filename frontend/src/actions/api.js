import axios from 'axios';

const baseUrl = "https://localhost:5001/api/";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
    product(url = baseUrl + 'Product/') {

        return {
            fetchAll: () => axios.get(url),
            create: newProduct => axios.post(url, newProduct),
            update: (id, updateProduct) => axios.put(url + id, updateProduct),
            delete: id => axios.delete(url + id),
            fetchById: id => axios.get(url + id)
        }
    }
};
