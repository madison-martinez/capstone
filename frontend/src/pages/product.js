import React from 'react';
import SingleProduct from '../components/SingleProduct';

function Product({ query }) {
    return (
        <div>
            <SingleProduct id={query.id} />
        </div>
    )
};
export default Product;