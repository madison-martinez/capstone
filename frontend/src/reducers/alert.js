import { alertActionTypes } from "../actions/alert";

export function alert(state = {}, action) {
  switch (action.type) {
    case alertActionTypes.SUCCESS:
      return {
        type: "ALERT_SUCCESS",
        message: action.message,
      };
    case alertActionTypes.ERROR:
      return {
        type: "ALERT_ERROR",
        message: action.message,
      };
    case alertActionTypes.RESET:
      return {};
    default:
      return state;
  }
};
