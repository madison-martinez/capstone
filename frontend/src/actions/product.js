import api from './api';


export const ACTION_TYPES = {
    CREATE: 'CREATE',
    UPDATE: 'UPDATE',
    DELETE: 'DELETE',
    FETCH_ALL: 'FETCH',
    FETCH_BY_ID: 'FETCH'
};

export const fetchAll = () => dispatch => {
    //GET REQUEST
    api.product().fetchAll()
    .then(
        response => {
        dispatch({
            type: ACTION_TYPES.FETCH_ALL,
            payload: response.data
        })
    })
    .catch(error => console.log(error)); //TODO SHOW ERRORS TO USER
};

export const fetchById = (id) => dispatch => {
    //GET REQUEST
    api.product().fetchById(id)
    .then(
        response => {
        dispatch({
            type: ACTION_TYPES.FETCH_BY_ID,
            payload: {...response.data, id}
        })
    })
    .catch(error => console.log(error)); //TODO SHOW ERRORS TO USER
};

export const create = ( data, onSuccess ) => dispatch => {
    //POST REQUEST
    api.product().create(data)
    .then(response => {
        dispatch({
            type: ACTION_TYPES.CREATE,
            payload: response.data
        })
        onSuccess();
    })
    .catch(error => console.log(error));
};

export const update = ( data, id, onSuccess ) => dispatch => {
    //PUT REQUEST
    api.product().update( data, id )
    .then(
        response => {
        dispatch({
            type: ACTION_TYPES.UPDATE,
            payload: { ...data, id }
        })
        onSuccess();
    })
    .catch(error => console.log(error));
};

export const deleteProduct = ( id, onSuccess ) => dispatch => {
    //DELETE REQUEST
    api.product().delete(id)
    .then(response => {
        dispatch({
            type: ACTION_TYPES.DELETE,
            payload: id
        })
        onSuccess();
    })
    .catch(error => console.log(error));
};