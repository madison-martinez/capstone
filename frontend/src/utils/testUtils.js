import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { reducers } from '../reducers';

function render(
  ui,
  {
    initialState,
    store = createStore(reducers, initialState ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

//mock middleware functions
const thunk = ({ dispatch, getState }) => next => action => {
    if (typeof action === 'function') {
      return action(dispatch, getState)
    }
    return next(action)
};

const fakeItem = () => ({
    id: '1',
    price: 20.00,
    title: 'Roses',
    description: 'Roses are Red',
});

export {
    fakeItem,
    thunk
};

export * from '@testing-library/react';
export { render };