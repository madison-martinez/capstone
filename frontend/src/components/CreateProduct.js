import React from 'react';
import { connect } from 'react-redux';
import Image from 'next/image';
import Form from './styles/FormStyles';
import * as actions from '../actions/product';
import useForm from '../utils/useForm';


const CreateProduct = (props) => {

    const { values, handleChange, setInitialForm, setValues } = useForm({
        title: '',
        description: '',
        price: 0,
        image: ''
    });

    const handleSubmit = e => {
        e.preventDefault();
        props.createProduct(values)
        setInitialForm();
    };

    const handleImageUpload = async e => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'farmersspecial');

        const res = await fetch(
            'https://api.cloudinary.com/v1_1/dzqeffkmp/image/upload',
            {
                method: 'POST',
                body: data
            });
        const file = await res.json();
        console.log(file)
        setValues({ image: file.secure_url });
    }

    return (
        <Form
            data-testid="form"
            onSubmit={handleSubmit}
        >
            <fieldset>
                <label htmlFor="file">
                    Image
                    <input
                        type="file"
                        id="file"
                        name="image"
                        placeholder="Upload an image"
                        onChange={handleImageUpload}
                    />
                </label>
                <label htmlFor="title">
                    Title
                    <input
                        type="text"
                        id="title"
                        name="title"
                        placeholder="Title"
                        required
                        value={values.title}
                        onChange={handleChange}
                    />
                </label>
                <label htmlFor="description">
                    Description
                    <textarea
                        id="description"
                        name="description"
                        placeholder="Description"
                        required
                        value={values.description}
                        onChange={handleChange}
                    />
                </label>
                <label htmlFor="price">
                    Price
                    <input
                        type="number"
                        id="price"
                        name="price"
                        placeholder="Price"
                        required
                        value={values.price}
                        onChange={handleChange}
                    />
                </label>
                <button type="submit">Submit</button>
            </fieldset>

            <Image
                src="/assets/photo-1560493676-04071c5f467b.jpeg"
                alt="fieldrows"
                layout="responsive"
                width={200}
                height={200}
            />

        </Form>
    )
};

const mapStateToProps = state => ({ productList: state.product.list });

const mapActionsToProps = {
    createProduct: actions.create,
};


export default connect(mapStateToProps, mapActionsToProps)(CreateProduct);