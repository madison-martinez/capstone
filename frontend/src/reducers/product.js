import { HYDRATE } from 'next-redux-wrapper';
import { ACTION_TYPES } from '../actions/product';

const initialState = {
    list: []
}

export const product = (state=initialState, action) => {
    switch (action.type) {
        case HYDRATE: 
            return { ...state, ...action.payload };
        case ACTION_TYPES.FETCH_ALL:
            return {
                ...state,
                list: [...action.payload]
            };
        case ACTION_TYPES.CREATE:
            return {
                ...state,
                list: [ ...state.list, action.payload ]
            };
        case ACTION_TYPES.UPDATE:
            return {
                ...state,
                list: state.list.map(x => 
                    x.id === action.payload.id ? action.payload : x)
            };
        case ACTION_TYPES.DELETE:
            return {
                ...state,
                list: state.list.filter(x =>
                    x.id !== action.payload )
            }
        case ACTION_TYPES.PRODUCT_SELECTED: 
            const selectedProduct = state.list.find(
                item => item.id === action.payload
            );
            return {
                ...state, 
                selectedProduct
            }
        default: 
            return state;
    }
}
