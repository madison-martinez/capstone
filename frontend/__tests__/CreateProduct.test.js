import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { rest } from "msw";
import { setupServer } from "msw/node";
import moxios from "moxios";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/react";
import CreateProduct from "../src/components/CreateProduct";
import { render, screen } from "../src/utils/testUtils";
import { ACTION_TYPES, create } from "../src/actions/product";
import { alertActionTypes } from "../src/actions/alert";

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  product: {
    list: [],
  },
};

const fakeNewItem = {
  id: 2,
  title: "fakeFood",
  description: "fakeYums",
  price: 2,
  image: "food.jpeg",
};

const fakeItemMissingFields = {
  id: 2,
  title: "",
  description: "",
  price: 2,
};

const mockSuccessFn = jest.fn();

const server = setupServer(
  rest.post(
    "https://localhost:5001/api/Product/:newProduct",
    (req, res, ctx) => {
      return res(ctx.json(fakeNewItem));
    }
  ),
  rest.post(
    "https://api.cloudinary.com/v1_1/dzqeffkmp/image/upload",
    (req, res, ctx) => {
      return res(ctx.status(200), ctx.json(fakeNewItem.image));
    }
  )
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  jest.clearAllMocks();
});
afterAll(() => server.close());

describe("<CreateProduct />", () => {
  let store;
  beforeEach(() => {
    moxios.install();
    store = mockStore(initialState);
  });
  afterEach(() => {
    moxios.uninstall();
  });

  it("should render empty form", () => {
    const { container } = render(<CreateProduct productList={[]} />);
    expect(container).toMatchSnapshot();
  });

  it("handles user inputting a new product", async () => {
    //render the form
    render(<CreateProduct store={store} />);
    expect(await screen.findByTestId("form")).toBeInTheDocument();
    //upload an image
    const file = new File(["hello"], fakeNewItem.image, { type: "image/png" });

    const input = await screen.findByTestId("image-element");
    userEvent.upload(input, file);
    expect(input.files[0]).toStrictEqual(file);
    expect(input.files).toHaveLength(1);

    //user types title, description, and price
    userEvent.type(screen.getByPlaceholderText("Title"), fakeNewItem.title);
    userEvent.type(
      screen.getByPlaceholderText("Description"),
      fakeNewItem.description
    );
    userEvent.type(
      screen.getByPlaceholderText("Price"),
      fakeNewItem.price.toString()
    );

    expect(screen.getByDisplayValue(fakeNewItem.title)).toBeInTheDocument();
    expect(screen.getByText(fakeNewItem.description)).toBeInTheDocument();
    //user clicks the submit button
    userEvent.click(screen.getByText("Submit"));
    //submit button fires dispatch to create product
    moxios.wait(function () {
      let request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: fakeNewItem,
      });
    });
    const expectedActions = [
      {
        type: ACTION_TYPES.CREATE,
        payload: fakeNewItem,
      },
    ];
    return store.dispatch(create(fakeNewItem, mockSuccessFn)).then(async () => {
      const actualAction = store.getActions();
      expect(actualAction).toEqual(expectedActions);
      await waitFor(() => expect(mockSuccessFn).toHaveBeenCalledTimes(1));
    });
  });

  it("checks that required fields are filled out", () => {
    render(<CreateProduct store={store} />);
    userEvent.click(screen.getByText("Submit"));
    //submit button fires dispatch to create product
    moxios.wait(function () {
      let request = moxios.requests.mostRecent();
      request.respondWith({
        status: 400,
        response: "Error: Request failed with status code 400",
      });
    });
    const expectedActions = [
      {
        type: alertActionTypes.ERROR,
        message: "Price must be between $1.00 and $10,000",
      },
      {
        type: alertActionTypes.ERROR,
        message: "Error: Request failed with status code 400",
      },
    ];
    return store
      .dispatch(create(fakeItemMissingFields, mockSuccessFn))
      .then(async () => {
        const actualAction = store.getActions();
        expect(actualAction).toEqual(expectedActions);
        await waitFor(() => expect(mockSuccessFn).toHaveBeenCalledTimes(0));
      });
  });
});
