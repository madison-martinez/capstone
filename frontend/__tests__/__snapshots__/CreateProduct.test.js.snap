// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct /> should render empty form 1`] = `
<div>
  <form
    class="FormStyles__Form-sc-1qtbzh2-0 cmkxQl"
    data-testid="form"
  >
    <fieldset>
      <label
        for="file"
      >
        Image
        <input
          data-testid="image-element"
          id="file"
          name="image"
          placeholder="Upload an image"
          type="file"
        />
      </label>
      <label
        for="title"
      >
        Title
        <input
          id="title"
          name="title"
          placeholder="Title"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="Description"
          required=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="Price"
          required=""
          type="number"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
    <img
      alt="fieldrows"
      height="500"
      layout="responsive"
      src="/assets/photo-1560493676-04071c5f467b.jpeg"
      width="500"
    />
  </form>
</div>
`;
